src/main/java/com/bootcamp/demo/dao/FirestoreDaoImpl.java

```diff
@@ -188,4 +188,23 @@ public Transaction addTransaction(Transaction transaction) {
         ApiFuture<WriteResult> document = firestoreDB.collection("transactions").document().set(transactionsData);
         return transaction;
     }
+
+    @Override
+    public String getPreferredCardId() {
+        CollectionReference ref = firestoreDB.collection(CARDS_COLLECTION);
+        Query stateQuery = ref.whereEqualTo("state", CardStateEnum.PREFERRED);
+        ApiFuture<QuerySnapshot> stateQuerySnapshot = stateQuery.get();
+        String cardId = "";
+        try {
+            if(stateQuerySnapshot.get().isEmpty()){
+                LOGGER.error("No preferred card set");
+                return "";
```

Comment:
If you want you could throw an exception (runtime exception - "bubble up") for this case that could be treated in the REST controller by responding with an appropriate message.